### PR TITLE
For lucas

### PR DESCRIPTION
**Pull request by asselitx**
Could you please review and comment? Only pay attention to the README-Prototype.md file. It is an update of what you saw on the other repo.

### CODE REVIEW

**Comment by asselitx** [comment]
@afishbeck and @timothyklemm could you please review? Only pay attention to the README-Prototype.md file. It is an update of what you saw on the other repo.

File: README-Prototype.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+The esdl tool aids with creating and managing ESDL-based and Dynamic ESDL services on an HPCC cluster. It consists of several different commands.
+
+## Overall notes for Tim and Tony
+
+Here is my overview of the design direction:
+
+1. Think of the manifest as config-as-code. The operator using the manifest to generate output shouldn't need to know which options to pass on the command line.
+2. Collapse the template file into the manifest.
+3. Think of the template as having two kinds of markup
```
Comment (afishbeck) [inline README-Prototype.md]: Picky, but after saying "collapse the template into the manifest" let's not refer to it as template anymore, it's the manifest.  My only real point being that templates sound like they are for more generic purposes than these manifests are.  These manifests describe actual service deployments and are only modified by providing variables/values.

File: README-Prototype.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    1. operations- Like `<include>` which will have an associated namespace and an explicit prefix.
+    2. boilerplate- Like `<Method>` which gives structure to the output. These elements will most often use the defaut 'no prefix'.
```
Comment (afishbeck) [inline README-Prototype.md]: "which describes the objects being configured"?

**Comment by afishbeck** [comment]
So here's an example of what I'm currently envisioning.  The final output format would be considered a semi internal detail, we are describing the objects being configured which are well known boilerplate (again in the future we can more object types and boilerplate, but we can also add a "build your own boilerplate".

```
<em:esdl-manifest xmlns:em="urn:hpcc:esdl:manifest" xmlns="urn:hpcc:esdl:bundle">
  <em:configuration>
    <ESDLBundle>
      <Service esdlService="WsFooBar" port="8899" name="MyWsFooBar">
        <Authorization/>
        <Scripts>
          <em:include file="/full/path/to/FooBarService-script1.xml"/>
          <em:include file="/full/path/to/FooBarService-script2.xml"/>
        </Scripts>
        <Methods>
          <Method name="FooBarReport" queryname="RoxieFooBarReport" querytype="roxie" url="https://roxievip:9876/roxie">
            <Scripts>
              <em:include file="/full/path/to/FooBarReport-script1.xml"/>
            </Scripts>
            <Gateways>
              <Gateway name="baz" url="www.baz.com/api"/>
            </Gateways>
          </Method>
        </Methods>
      </Service>
      <Definitions>
        <em:include file="/full/path/to/ws_foobar.ecm" service="WsFooBar" searchPaths="/path/one:/path/one/two"/>
      </Definitions>
    </ESDLBundle>
  </em:configuration>
</em:esdl-manifest>
```

**Comment by afishbeck** [comment]
I am just describing what I want to add, not worrying about what the internal file looks like.  We will need to play with how to generate the internal ids that esdl_binding will like, or maybe add support in esdl_binding to figure those out on it's own since everything here is completely self contained.

**Comment by afishbeck** [comment]
I can help when you get to that point.

Notice the Authentication tag.  When we have new things to describe we can figure out where they best fit.  This is where the auth_features would come from.

But I think the internal boilerplate "template" to get from here to a bundle or binding file is pretty straight forward.

**Comment by afishbeck** [comment]
Further in the future we can make the bundle output look more like this.  The existing formats are really internal details based on what was stored in DALI.  They could use refactoring for modern purposes.

File: README-Prototype.md (hunk 2 above)
Comment (asselitx) [reply README-Prototype.md]: OK yeah your wording makes more sense.

File: README-Prototype.md (hunk 1 above)
Comment (asselitx) [reply README-Prototype.md]: Ok yep I missed those references.

**Comment by asselitx** [comment]
OK Tony so you're example:

> 
> ```
> <em:esdl-manifest xmlns:em="urn:hpcc:esdl:manifest" xmlns="urn:hpcc:esdl:bundle">
>   <em:configuration>
>     <ESDLBundle>
>       <Service esdlService="WsFooBar" port="8899" name="MyWsFooBar">
>         <Authorization/>
> ```
> ...

does also look very close to some suggestions Tim made. Do you imagine the default namespace (eg urn:hpcc:esdl:bundle) would change based on output (bundle vs binding vs fun new stuff)? Should that namespace be the primary driver of 'state' in the tool for customizing parsing and output behavior?

**Comment by timothyklemm** [comment]
I had a couple thoughts about the example, besides noting the similarities Terrence mentioned.

I would think that any element that doesn't map directly into the output (at least relative to its parent) should be considered our markup in our namespace. Mapping the example to what we currently expect in the binding, several of the elements with no namespace prefix should be prefixed. If we're going to redefine the expected format to more closely match the example, then I could be wrong.

The <Methods> element seems extraneous in the example. Either we're redefining the expected format and don't need to retain it, or we're already including the capability to rearrange the output and can create it as needed.

**Comment by afishbeck** [comment]
@asselitx yes, the namespace of the tag below configuration will tell you which type of boilerplate transform you are performing.  SO in the case above ESDLBundle has no prefix so has the defuault namespace.  That namespace is "urn:hpcc:esdl:bundle" so you know you are processing a bundle.  Of course the tag itself could be your clue.  Since the root tag of the "urn:hpcc:esdl:bundle" schema will be "ESDLBundle".

You can think of your boilerplate code for that namespace as the real template which is why I didn't want to call the manifest a template.

**Comment by timothyklemm** [comment]
If I remember correctly, DESDL documentation (the official kind) instructs users to add any XML they need to send to the roxie within Method elements. The manifest must, therefore, accept free-form XML that is unknown to us and this content is unlikely to be prefixed. I'm still thinking that if it's an element we will look for (and make decisions about) it needs a namespace AND prefix, leaving un-prefixed markup as content that we propagate directly into the output.

**Comment by afishbeck** [comment]
I see.  Well we can support free form stuff to send.  We should probably wrap it in something though, even if for now we output it in the method tag.  That free form content would not have a namespace that we know or care about.  We can think about it further, and don't need to support it this first release.  That may be the place that we support CDATA in the manifest.. but I'm not sure yet.

**Pull request merged**